### PR TITLE
SVMSGD class added

### PR DESCRIPTION
**Pull request by joaofaro**
SVMSGD class added.

### CODE REVIEW

File: modules/ml/src/svmsgd.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    _nIterations = nIterations;  // number of training iterations
+
+    // True only in the first predict iteration
+    _initPredict = true;
+
+    // Online update flag
+    _onlineUpdate = true;
+
+    // Learn rate decay: _learnRate = _learnRate * _learnDecay
+    _learnRateDecay = 0.1;
```
Comment (berak) [inline modules/ml/src/svmsgd.cpp]: _learnRateDecay = 0.1f; (float truncation)

File: modules/ml/src/svmsgd.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    innerProduct = calcInnerProduct(newFeature.ptr<float>(0));
+    
+    // Resultant label (-1 or 1)
+    int label = (innerProduct>=0) ? 1 : -1;
+
+    if (_onlineUpdate){
+        // Update the featuresSlider with newFeature and _labelSlider with label
+        newFeature.row(0).copyTo(_featuresSlider.row(_sliderCounter));
+        _labelSlider[_sliderCounter] = label;
```
Comment (berak) [inline modules/ml/src/svmsgd.cpp]: _labelSlider[_sliderCounter] = float(label);

File: modules/ml/src/svmsgd.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+        // Update weights with a random index
+        if (_sliderCounter == _slidingWindowSize-1){
+            generateRandomIndex();
+            updateWeights(innerProduct, _featuresSlider.ptr<float>(_randomIndex), _labelSlider[_randomIndex]);
```
Comment (berak) [inline modules/ml/src/svmsgd.cpp]: int(_labelSlider[_randomIndex])   

(reverse of above)

File: modules/ml/src/svmsgd.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        }
+
+        // _sliderCounter++ if < _slidingWindowSize
+        _sliderCounter = (_sliderCounter == _slidingWindowSize-1) ? 0 : (_sliderCounter+1);
+    }
+
+    return label;
```
Comment (berak) [inline modules/ml/src/svmsgd.cpp]: return float(label);

File: doc/opencv.bib [hunk 5]
```diff
@@ -848,3 +848,12 @@ @incollection{Liao2007
...
 }
+@incollection{bottou2010large,
+  title={Large-scale machine learning with stochastic gradient descent},
+  author={Bottou, L{\'e}on},
+  booktitle={Proceedings of COMPSTAT'2010},
+  pages={177--186},
+  year={2010},
+  publisher={Springer}
+}
+
```
Comment (berak) [inline doc/opencv.bib]: please remove blank line at end

File: modules/ml/src/svmsgd.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        _nFeatures = newFeature.cols;
+        _slidingWindowSize = _predictSlidingWindowSize;
+        _featuresSlider = cv::Mat::zeros(_slidingWindowSize, _nFeatures, CV_32F);
+        _initPredict = false;
+        _labelSlider = new float[_predictSlidingWindowSize]();
+        _learnRate = _learnRate * _learnRateDecay;
+    }
+
+    innerProduct = calcInnerProduct(newFeature.ptr<float>(0));
+
```
Comment (berak) [inline modules/ml/src/svmsgd.cpp]: please remove 4 whitespaces in 151

File: modules/ml/src/svmsgd.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        _sliderCounter = (_sliderCounter == _slidingWindowSize-1) ? 0 : (_sliderCounter+1);
+    }
+
+    return label;
+}
+
+void SVMSGD::generateRandomIndex(){
+    // Choose random sample, using Mikolov's fast almost-uniform random number
+    _randomNumber = _randomNumber * (unsigned long long) 25214903917 + 11;
+    _randomIndex = _randomNumber % (unsigned long long) _slidingWindowSize;
```
Comment (berak) [inline modules/ml/src/svmsgd.cpp]: _randomIndex is a 'unsigned int' , so too small for the assigned 'unsigned long long'

File: modules/ml/src/svmsgd.cpp (hunk 1 above)
Comment (joaofaro) [inline modules/ml/src/svmsgd.cpp]: _learnRateDecay will be defined by an input variable, like _predictSlidingWindowSize

**Comment by sturkmen72** [comment]
you can check compiled help file http://pullrequest.opencv.org/buildbot/export/pr/5291/docs/

**Comment by joaofaro** [comment]
Ok, I got stuck now. I couldn't find the doc of SVMSGD, so I made a search through the classes listed on the ml module and I found theirs description on  ml.hpp. Should I add the new class as an inherited one from StatModel?

File: modules/ml/include/opencv2/ml.hpp [hunk 8]
```diff
@@ -1488,6 +1488,127 @@ CV_EXPORTS void randGaussMixture( InputArray means, InputArray covs, InputArray
...
+
+First, create the SVMSGD object. To enable the online update, a value for updateFrequency should be defined.
+
+Then the SVM model can be trained using the train features and the correspondent labels.
+
+After that, the label of a new feature vector can be predicted using the predict function. If the updateFrequency was defined in the constructor, the predict function will update the weights automatically.
+
+@code
+// Initialize object
+SvmSgd SVMSGD;
```
Comment (sturkmen72) [inline modules/ml/include/opencv2/ml.hpp]: isn't it like SVMSGD SvmSgd;

Comment (joaofaro) [inline modules/ml/include/opencv2/ml.hpp]: Yes it is! Thanks for pointing the error!

**Comment by vpisarev** [comment]
@joaofaro, thank you very much for the contribution and big sorry for delay!
it looks definitely useful, but the code style is quite different from all the existing ml classifiers. As you can see, we use "open interface - hidden implementation" pattern for all of them, it means that the class constructors, as well as all the members and internal methods are put into the implementation class, whereas the interface is represented by the base abstract class without constructor. If you have time, please, refactor your class to follow this pattern

**Comment by vpisarev** [comment]
Also, it would be useful to have some example on how to use this algorithm

**Comment by alalek** [comment]
Improved in #6096

**Comment by joaofaro** [comment]
Thank you very much! As I said to Marina, I recently moved to another country and got a new job, so I have been very busy lately. I was planning to do the refactoring later but I I am glad to see that you decided to make it!

**Pull request closed without merging**